src/daemon/pex/pex.go

```diff
@@ -135,545 +131,237 @@ func (peer *Peer) String() string {
-// return all exchangeable public peers if count is 0.
-func (pl *Peerlist) RandomExchgPublic(count int) []Peer {
-	pl.lock.RLock()
-	defer pl.lock.RUnlock()
-
-	return pl.randomExchg(count, false)
-}
-
-// RandomExchgAll returns n random exchangeable peers, including private peers.
-// return all exchangeable peers if count is 0.
-func (pl *Peerlist) RandomExchgAll(count int) []Peer {
-	pl.lock.RLock()
-	defer pl.lock.RUnlock()
-
-	return pl.randomExchg(count, true)
-}
-
-// RandomPublic returns n random peers, or all of the peers, whichever is lower.
-// If count is 0, all of the peers are returned, shuffled.  Will not include
-// private peers.
-func (pl *Peerlist) RandomPublic(count int) []Peer {
-	pl.lock.RLock()
-	defer pl.lock.RUnlock()
-
-	return pl.random(count, false)
-}
-
-// RandomAll returns n random peers, or all of the peers, whichever is lower.
-// If count is 0, all of the peers are returned, shuffled.  Includes private
-// peers.
-func (pl *Peerlist) RandomAll(count int) []Peer {
-	pl.lock.RLock()
-	defer pl.lock.RUnlock()
-
-	return pl.random(count, true)
-}
-
-// PeerJSON is for saving and loading peers to disk. Some fields are strange,
-// to be backwards compatible due to variable name changes
-type PeerJSON struct {
-	Addr string // An address of the form ip:port
-	// Unix timestamp when this peer was last seen.
-	// This could be a time.Time string or an int64 timestamp
-	LastSeen        interface{}
-	Private         bool  // Whether it should omitted from public requests
-	Trusted         bool  // Whether this peer is trusted
-	HasIncomePort   *bool `json:"HasIncomePort,omitempty"` // Whether this peer has incoming port [DEPRECATED]
-	HasIncomingPort *bool // Whether this peer has incoming port
-}
-
-// NewPeerJSON returns a PeerJSON from a Peer
-func NewPeerJSON(p Peer) PeerJSON {
-	hasIncomingPort := p.HasIncomingPort
-
-	return PeerJSON{
-		Addr:            p.Addr,
-		LastSeen:        p.LastSeen,
-		Private:         p.Private,
-		Trusted:         p.Trusted,
-		HasIncomingPort: &hasIncomingPort,
+	// file does not exist
+	if peers == nil {
+		return nil
 	}
-}
 
-// NewPeerFromJSON converts a PeerJSON to a Peer
-func NewPeerFromJSON(p PeerJSON) (Peer, error) {
-	hasIncomingPort := false
-	if p.HasIncomingPort != nil {
-		hasIncomingPort = *p.HasIncomingPort
-	} else if p.HasIncomePort != nil {
-		hasIncomingPort = *p.HasIncomePort
+	// remove invalid peers and limit the max number of peers to pex.Config.Max
+	l := len(peers)
+	if l > px.Config.Max {
+		l = px.Config.Max
 	}
 
-	// LastSeen could be a RFC3339Nano timestamp or an int64 unix timestamp
-	var lastSeen int64
-	switch p.LastSeen.(type) {
-	case string:
-		lastSeenTime, err := time.Parse(time.RFC3339Nano, p.LastSeen.(string))
-		if err != nil {
-			return Peer{}, err
+	var validPeers []Peer
+	for addr, p := range peers {
+		if validateAddress(addr, px.Config.AllowLocalhost) {
+			validPeers = append(validPeers, *p)
+			l--
```

Comment:
why not `if len(validPeers) >= px.Config.Max { break }`

Reply:
or process all then `if len(validPeers) > px.Config.Max { validPeers = validPeers[:px.Config.Max] }` after the loop